File: oiioii-sdk/gptmail.js
const https = require('https');
const { createProxyAgent } = require('./rotating_proxy');

const GPTMAIL_HOST = 'mail.chatgpt.org.uk';
const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36';

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

// --- 底层 HTTPS 请求 ---

function gptmailRequest(path, options = {}, agent = null) {
  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: GPTMAIL_HOST,
      path,
      method: options.method || 'GET',
      agent: agent || undefined,
      timeout: options.timeout || 15000,
      headers: {
        'User-Agent': UA,
        ...options.headers
      }
    }, (res) => {
      const cookies = (res.headers['set-cookie'] || []).map(c => c.split(';')[0]);
      const chunks = [];
      res.on('data', d => chunks.push(d));
      res.on('end', () => {
        const text = Buffer.concat(chunks).toString();
        let json = null;
        try { json = JSON.parse(text); } catch (e) {}
        resolve({ status: res.statusCode, data: text, json, cookies });
      });
    });
    req.on('error', reject);
    req.on('timeout', () => { req.destroy(new Error(`GPTMail request timeout after ${options.timeout || 15000}ms`)); });
    if (options.body) req.write(options.body);
    req.end();
  });
}

// --- 随机前缀 ---

function randomPrefix() {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let s = '';
  for (let i = 0; i < 8 + Math.floor(Math.random() * 4); i++) {
    s += chars[Math.floor(Math.random() * chars.length)];
  }
  return s;
}

// --- 域名池（带 30 分钟缓存） ---

const FALLBACK_DOMAIN = 'ppoo.ccwu.cc';
const DOMAIN_TTL = 30 * 60 * 1000;
let _domainCache = null;
let _domainCacheAt = 0;

async function getValidDomains() {
  const now = Date.now();
  if (_domainCache && now - _domainCacheAt < DOMAIN_TTL) return _domainCache;

  const maxRetries = 3;
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      const res = await gptmailRequest('/api/domains/status', {
        headers: { 'Accept': 'application/json' },
        timeout: 20000
      });
      const list = (res.json?.data?.domains || [])
        .filter(d => d.mx_valid && d.is_active)
        .map(d => d.domain_name);
      if (list.length > 0) {
        _domainCache = list;
        _domainCacheAt = now;
        console.log(`  [domains] refreshed pool: ${list.length} valid domains`);
        return _domainCache;
      }
      throw new Error('empty domain list');
    } catch (e) {
      if (attempt < maxRetries - 1) {
        await sleep(1000 + attempt * 1000);
        continue;
      }
      console.log(`  [warn] fetch domains failed: ${e.message}, using cached/fallback`);
    }
  }
  return _domainCache || [FALLBACK_DOMAIN];
}

// --- 生成随机临时邮箱 ---

async function generateEmail() {
  console.log('[1/5] Generating random email...');
  const prefix = randomPrefix();
  const domains = await getValidDomains();
  const domain = domains[Math.floor(Math.random() * domains.length)];
  const email = `${prefix}@${domain}`;
  console.log(`  Email: ${email}  (domain pool: ${domains.length})`);
  return email;
}

// --- 建立 GPTMail session ---

async function getFullBrowserCookies(email, agent = null) {
  const maxRetries = 5;
  for (let attempt = 0; attempt < maxRetries; attempt++) {
    try {
      const pageRes = await gptmailRequest(`/zh/${email}`, {
        headers: { 'Accept': 'text/html,application/xhtml+xml', 'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8' }
      }, agent);

      const cookies = pageRes.cookies || [];
      const cookieStr = cookies.join('; ');
      const payload = JSON.stringify({ email });

      const tokenRes = await gptmailRequest('/api/inbox-token', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(payload),
          'Cookie': cookieStr,
          'Referer': `https://${GPTMAIL_HOST}/zh/${email}`
        },
        body: payload
      }, agent);

      const json = tokenRes.json;
      if (json?.success && json?.auth?.token) {
        const result = cookies;
        result._inboxToken = json.auth.token;
        return result;
      }

      if (tokenRes.data && tokenRes.data.includes('Too many requests')) {
        if (attempt < maxRetries - 1) {
          await sleep(2000 + attempt * 2000);
          continue;
        }
      }
      throw new Error(`inbox-token failed: ${JSON.stringify(json || tokenRes.data).substring(0, 100)}`);
    } catch (e) {
      if (attempt === maxRetries - 1) throw e;
      await sleep(2000 + attempt * 2000);
    }
  }
  throw new Error('inbox-token failed after retries');
}

// --- 查询收件箱 ---

async function getEmailsWithCookies(email, allCookies, agent = null) {
  const cookieStr = allCookies.filter(c => c.startsWith('gm_sid')).join('; ');
  const res = await gptmailRequest(`/api/emails?email=${encodeURIComponent(email)}`, {
    headers: {
      'Cookie': cookieStr,
      'X-Inbox-Token': allCookies._inboxToken || '',
      'Accept': 'application/json',
      'Referer': `https://${GPTMAIL_HOST}/zh/${email}`
    }
  }, agent);
  return res.json || { success: false, error: (res.data || '').substring(0, 100) };
}

// --- 轮询收件箱并提取验证码 ---

async function waitForCode(email, maxWait = 90000, agent = null, preCookies = null) {
  console.log('\n[4/5] Waiting for verification code...');

  let allCookies = preCookies || await getFullBrowserCookies(email, agent);
  console.log(`  Got ${allCookies.length} cookies${preCookies ? ' (prebuilt)' : ''}`);

  const startTime = Date.now();
  let emptyCount = 0;
  const REBUILD_THRESHOLD = 10;

  while (Date.now() - startTime < maxWait) {
    const result = await getEmailsWithCookies(email, allCookies, agent);

    if (result?.success && result.data?.emails?.length > 0) {
      const mail = result.data.emails[0];
      console.log(`\n  Got email from: ${mail.from_address}`);
      console.log(`  Subject: ${mail.subject}`);

      const content = mail.content || mail.subject || '';
      const codeMatch = content.match(/\b(\d{4,6})\b/);
      if (codeMatch) {
        console.log(`  Verification code: ${codeMatch[1]}`);
        return codeMatch[1];
      }

      const subjMatch = (mail.subject || '').match(/(\d{4,6})/);
      if (subjMatch) {
        console.log(`  Verification code: ${subjMatch[1]}`);
        return subjMatch[1];
      }

      throw new Error('Could not extract verification code from email');
    }

    emptyCount++;

    if (emptyCount === REBUILD_THRESHOLD) {
      const elapsed = ((Date.now() - startTime) / 1000).toFixed(0);
      console.log(`\n  [session rebuild] ${emptyCount} empty polls in ${elapsed}s, rebuilding with fresh agent...`);
      try {
        if (agent) agent.destroy();
        agent = createProxyAgent();
        allCookies = await getFullBrowserCookies(email, agent);
        console.log(`  [session rebuild] OK, got ${allCookies.length} cookies (new IP)`);
        emptyCount = 0;
      } catch (e) {
        console.log(`  [session rebuild] failed: ${e.message}, continuing with old session`);
      }
    }

    const elapsed = ((Date.now() - startTime) / 1000).toFixed(0);
    process.stdout.write(`  Polling... (${elapsed}s)\r`);
    await sleep(1500);
  }

  throw new Error(`Timeout: no email received within ${maxWait / 1000}s`);
}

// --- 简化版直连接口（无代理） ---

function httpsRequestDirect(path, options = {}) {
  return new Promise((resolve, reject) => {
    const req = https.request({
      hostname: GPTMAIL_HOST,
      port: 443,
      path,
      method: options.method || 'GET',
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'application/json',
        ...options.headers
      }
    }, (res) => {
      const cookies = res.headers['set-cookie'];
      const chunks = [];
      res.on('data', chunk => chunks.push(chunk));
      res.on('end', () => {
        const text = Buffer.concat(chunks).toString('utf-8');
        let json;
        try { json = JSON.parse(text); } catch (e) { json = null; }
        resolve({ status: res.statusCode, data: text, json, headers: res.headers, cookies });
      });
    });
    req.on('error', reject);
    if (options.body) req.write(options.body);
    req.end();
  });
}

async function getGptMailSession(email) {
  const pageRes = await httpsRequestDirect(`/zh/${email}`, {
    headers: { 'Accept': 'text/html' }
  });

  if (!pageRes.cookies || pageRes.cookies.length === 0) {
    throw new Error('GPTMail: no session cookie returned');
  }

  const cookie = pageRes.cookies[0].split(';')[0];

  const payload = JSON.stringify({ email });
  const tokenRes = await httpsRequestDirect('/api/inbox-token', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': String(Buffer.byteLength(payload)),
      'Cookie': cookie
    },
    body: payload
  });

  if (!tokenRes.json?.success || !tokenRes.json?.auth?.token) {
    throw new Error(`GPTMail token failed: ${tokenRes.data.substring(0, 200)}`);
  }

  return {
    cookie,
    token: tokenRes.json.auth.token,
    email
  };
}

async function getEmails(session) {
  const res = await httpsRequestDirect(`/api/emails?email=${encodeURIComponent(session.email)}`, {
    headers: { 'Cookie': session.cookie }
  });
  return res.json;
}

async function getEmailDetail(session, emailId) {
  const res = await httpsRequestDirect(`/api/email/${emailId}`, {
    headers: { 'Cookie': session.cookie }
  });
  return res.json;
}

module.exports = {
  generateEmail,
  getValidDomains,
  getFullBrowserCookies,
  getEmailsWithCookies,
  waitForCode,
  getGptMailSession,
  getEmails,
  getEmailDetail,
  gptmailRequest,
  sleep
};

File: insmind-sdk/insmind/gptmail.py
"""GPTMail 临时邮箱。"""

from __future__ import annotations

import random
import re
import string
import time
import urllib.parse
from typing import Callable, Optional

from . import constants as C
from .http import request


class GPTMailClient:
    _domain_cache: list[str] | None = None
    _domain_cache_at = 0.0

    def __init__(self) -> None:
        self.email = ""
        self.cookies: list[str] = []
        self.token = ""

    @classmethod
    def valid_domains(cls) -> list[str]:
        now = time.time()
        if cls._domain_cache and now - cls._domain_cache_at < 1800:
            return cls._domain_cache
        result = request(
            f"https://{C.GPTMAIL_HOST}/api/domains/status",
            headers={"Accept": "application/json"},
            timeout=20,
        )
        domains = [
            item.get("domain_name", "")
            for item in (((result.get("json") or {}).get("data") or {}).get("domains") or [])
            if item.get("mx_valid") and item.get("is_active") and item.get("domain_name")
        ]
        if domains:
            cls._domain_cache = domains
            cls._domain_cache_at = now
            return domains
        return cls._domain_cache or [C.GPTMAIL_FALLBACK_DOMAIN]

    @staticmethod
    def random_prefix() -> str:
        chars = string.ascii_lowercase + string.digits
        return "".join(random.choice(chars) for _ in range(random.randint(8, 11)))

    def generate_email(self, prefer_fallback: bool = True) -> str:
        domain = C.GPTMAIL_FALLBACK_DOMAIN if prefer_fallback else random.choice(self.valid_domains())
        self.email = f"{self.random_prefix()}@{domain}"
        return self.email

    def build_session(self, email: Optional[str] = None) -> dict:
        self.email = (email or self.email).strip()
        if not self.email:
            raise ValueError("email is required")
        request(
            f"https://{C.GPTMAIL_HOST}/zh/{urllib.parse.quote(self.email)}",
            headers={"Accept": "text/html,application/xhtml+xml", "Accept-Language": "zh-CN,zh;q=0.9"},
            timeout=20,
        )
        token_result = request(
            f"https://{C.GPTMAIL_HOST}/api/inbox-token",
            method="POST",
            headers={
                "Accept": "application/json",
                "Content-Type": "application/json",
                "Referer": f"https://{C.GPTMAIL_HOST}/zh/{self.email}",
            },
            body={"email": self.email},
            timeout=20,
        )
        data = token_result.get("json") or {}
        self.token = (((data.get("auth") or {}).get("token")) if data.get("success") else "") or ""
        self.cookies = token_result.get("cookies") or []
        if not self.token:
            raise RuntimeError(f"GPTMail inbox-token failed: {token_result.get('text', '')[:200]}")
        if not any(c.startswith("gm_sid=") for c in self.cookies):
            raise RuntimeError(f"GPTMail missing gm_sid cookie: {self.cookies}")
        return {"email": self.email, "cookies": self.cookies, "token": self.token}

    def _cookie_header(self) -> str:
        gm = [c for c in self.cookies if c.startswith("gm_sid=")]
        return "; ".join(gm or self.cookies)

    def list_emails(self) -> list[dict]:
        if not self.token or not self.cookies:
            self.build_session(self.email)
        result = request(
            f"https://{C.GPTMAIL_HOST}/api/emails?email={urllib.parse.quote(self.email)}",
            headers={
                "Accept": "application/json",
                "Cookie": self._cookie_header(),
                "X-Inbox-Token": self.token,
                "Referer": f"https://{C.GPTMAIL_HOST}/zh/{self.email}",
            },
            timeout=20,
        )
        if result.get("status") in (401, 403):
            self.build_session(self.email)
            result = request(
                f"https://{C.GPTMAIL_HOST}/api/emails?email={urllib.parse.quote(self.email)}",
                headers={
                    "Accept": "application/json",
                    "Cookie": self._cookie_header(),
                    "X-Inbox-Token": self.token,
                    "Referer": f"https://{C.GPTMAIL_HOST}/zh/{self.email}",
                },
                timeout=20,
            )
        data = result.get("json") or {}
        if not data.get("success"):
            return []
        return ((data.get("data") or {}).get("emails") or [])

    def email_detail(self, message: dict) -> dict:
        message_id = message.get("id") or message.get("email_id") or message.get("_id")
        if not message_id:
            return message
        result = request(
            f"https://{C.GPTMAIL_HOST}/api/email/{urllib.parse.quote(str(message_id))}",
            headers={
                "Accept": "application/json",
                "Cookie": self._cookie_header(),
                "X-Inbox-Token": self.token,
            },
            timeout=20,
        )
        data = result.get("json") or {}
        detail = data.get("data") if data.get("success") else None
        return {**message, **detail} if isinstance(detail, dict) else message

    @staticmethod
    def extract_code(text: str, length: int = 6) -> str:
        if not text:
            return ""
        compact = re.sub(r"(?is)<style.*?</style>", " ", text)
        compact = re.sub(r"(?is)<script.*?</script>", " ", compact)
        compact = re.sub(r"(?s)<[^>]+>", " ", compact)
        compact = re.sub(r"&nbsp;", " ", compact)
        compact = re.sub(r"\s+", " ", compact)
        match = re.search(rf"\b(\d{{{length}}})\b", compact)
        return match.group(1) if match else ""

    def wait_for_code(
        self,
        max_wait: int = 90,
        expected_length: int = 6,
        message_filter: Optional[Callable[[dict], bool]] = None,
    ) -> str:
        deadline = time.time() + max_wait
        while time.time() < deadline:
            for message in self.list_emails():
                detail = self.email_detail(message)
                if message_filter and not message_filter(detail):
                    continue
                body = " ".join(
                    str(detail.get(key) or "")
                    for key in ("subject", "content", "text_content", "html_content")
                )
                lower = body.lower()
                if "insmind" in lower or "verify your email" in lower or "验证码" in body or not message_filter:
                    code = self.extract_code(body, expected_length)
                    if code:
                        return code
            time.sleep(2)
        raise TimeoutError(f"no verification email within {max_wait}s")
